server/service/src/rnr_form/update.rs

```diff
@@ -0,0 +1,199 @@
+use crate::{activity_log::activity_log_entry, service_provider::ServiceContext};
+
+use repository::{
+    ActivityLogType, RepositoryError, RnRForm, RnRFormLineRow, RnRFormLineRowRepository,
+    RnRFormStatus,
+};
+
+use super::{query::get_rnr_form, validate::check_rnr_form_exists};
+
+#[derive(Default, Debug, PartialEq, Clone)]
+pub struct UpdateRnRFormLine {
+    pub id: String,
+    pub quantity_received: Option<f64>,
+    pub quantity_consumed: Option<f64>,
+    pub adjustments: Option<f64>,
+    pub stock_out_duration: i32,
+    pub adjusted_quantity_consumed: f64,
+    pub final_balance: f64,
+    pub maximum_quantity: f64,
+    pub requested_quantity: f64,
+    pub comment: Option<String>,
+    pub confirmed: bool,
+}
+
+#[derive(Default, Debug, PartialEq, Clone)]
+pub struct UpdateRnRForm {
+    pub id: String,
+    pub lines: Vec<UpdateRnRFormLine>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UpdateRnRFormError {
+    DatabaseError(RepositoryError),
+    InternalError(String),
+    RnRFormDoesNotExist,
+    RnRFormAlreadyFinalised,
+    UpdatedRnRFormDoesNotExist,
+    LineError {
+        line_id: String,
+        error: UpdateRnRFormLineError,
+    },
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UpdateRnRFormLineError {
+    LineDoesNotExist,
+    LineDoesNotBelongToRnRForm,
+    ValuesDoNotBalance,
+    CannotRequestNegativeQuantity,
+}
+
+pub fn update_rnr_form(
+    ctx: &ServiceContext,
+    store_id: &str,
+    input: UpdateRnRForm,
+) -> Result<RnRForm, UpdateRnRFormError> {
+    let rnr_form = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let line_data = validate(ctx, &input)?;
+            let rnr_form_lines = generate(line_data);
+
+            let rnr_form_line_repo = RnRFormLineRowRepository::new(connection);
+
+            for line in rnr_form_lines {
+                rnr_form_line_repo.upsert_one(&line)?;
+            }
+
+            activity_log_entry(
+                ctx,
+                ActivityLogType::RnrFormUpdated,
+                Some(input.id.clone()),
+                None,
```

Comment:
thought for later: adding in some detail, so the user can see what the change was.
this is a can o worms, which is why I think later!